stmhal/sdcard.c

```diff
@@ -346,22 +358,82 @@ STATIC mp_obj_t sd_write(mp_obj_t self, mp_obj_t block_num, mp_obj_t data) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_3(sd_write_obj, sd_write);
 
-STATIC const mp_map_elem_t sdcard_locals_dict_table[] = {
+STATIC mp_obj_t pyb_sdcard_readblocks(mp_obj_t self, mp_obj_t block_num, mp_obj_t buf) {
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_WRITE);
+    mp_uint_t ret = sdcard_read_blocks(bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / SDCARD_BLOCK_SIZE);
+    return mp_obj_new_bool(ret == 0);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(pyb_sdcard_readblocks_obj, pyb_sdcard_readblocks);
+
+STATIC mp_obj_t pyb_sdcard_writeblocks(mp_obj_t self, mp_obj_t block_num, mp_obj_t buf) {
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_READ);
+    mp_uint_t ret = sdcard_write_blocks(bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / SDCARD_BLOCK_SIZE);
+    return mp_obj_new_bool(ret == 0);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(pyb_sdcard_writeblocks_obj, pyb_sdcard_writeblocks);
+
+STATIC mp_obj_t pyb_sdcard_ioctl(mp_obj_t self, mp_obj_t cmd_in, mp_obj_t arg_in) {
+    mp_int_t cmd = mp_obj_get_int(cmd_in);
+    switch (cmd) {
+        case 1: // INIT
+            if (!sdcard_power_on()) {
+                return MP_OBJ_NEW_SMALL_INT(-1); // error
+            }
+            return MP_OBJ_NEW_SMALL_INT(0); // success
+
+        case 2: // DEINIT
+            sdcard_power_off();
+            return MP_OBJ_NEW_SMALL_INT(0); // success
+
+        case 3: // SYNC
+            // nothing to do
+            return MP_OBJ_NEW_SMALL_INT(0); // success
+
+        case 4: // SEC_COUNT
+            return MP_OBJ_NEW_SMALL_INT(0); // TODO
+
+        case 5: // SEC_SIZE
```

Comment:
> True.  But they need to be moved out of stmhal/diskio.c and into a
> header.  I guess extmod/fsusermount.h?

Yes, there, if you don't feel like making a bdev.h in some dir (extmod
is good enough).